ibis/backends/clickhouse/__init__.py

```diff
@@ -184,26 +275,136 @@ def _client_execute(self, query, external_tables=None):
             )
 
+    def to_pyarrow_batches(
+        self,
+        expr: ir.Expr,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        limit: int | str | None = None,
+        chunk_size: int = 1_000_000,
+        **kwargs: Any,
+    ) -> pa.RecordBatchReader:
+        """Execute expression and return results in an iterator of pyarrow
+        record batches.
+
+        This method is eager and will execute the associated expression
+        immediately.
+
+        Parameters
+        ----------
+        expr
+            Ibis expression to export to pyarrow
+        limit
+            An integer to effect a specific row limit. A value of `None` means
+            "no limit". The default is in `ibis/config.py`.
+        params
+            Mapping of scalar parameter expressions to value.
+        chunk_size
+            Number of rows in each returned record batch.
+
+        Returns
+        -------
+        results
+            RecordBatchReader
+        """
+        pa = self._import_pyarrow()
+
+        from ibis.backends.pyarrow.datatypes import ibis_to_pyarrow_struct
+
+        schema = self._table_or_column_schema(expr)
+
+        def _batches():
+            for batch in self._cursor_batches(
+                expr, params=params, limit=limit, chunk_size=chunk_size
+            ):
+                struct_array = pa.array(
+                    map(tuple, batch),
+                    type=ibis_to_pyarrow_struct(schema),
+                )
+                yield pa.RecordBatch.from_struct_array(struct_array)
+
+        return pa.RecordBatchReader.from_batches(schema.to_pyarrow(), _batches())
+
     def _cursor_batches(
         self,
         expr: ir.Expr,
         params: Mapping[ir.Scalar, Any] | None = None,
         limit: int | str | None = None,
         chunk_size: int = 1_000_000,
     ) -> Iterable[list]:
-        query_ast = self.compiler.to_ast_ensure_limit(expr, limit, params=params)
-        sql = query_ast.compile()
-        cursor = self.raw_sql(str(sql))
+        sql = self.compile(expr, limit=limit, params=params)
+        cursor = self.raw_sql(sql)
         try:
             while batch := cursor.fetchmany(chunk_size):
                 yield batch
         finally:
             cursor.close()
 
+    def execute(
+        self,
+        expr: ir.Expr,
+        limit: str | None = 'default',
+        external_tables: Mapping[str, pd.DataFrame] | None = None,
+        **kwargs: Any,
+    ) -> Any:
+        """Execute an expression."""
+        table_expr = expr.as_table()
+        sql = self.compile(table_expr, limit=limit, **kwargs)
+        self._log(sql)
+
+        for memtable in graph.traverse(an.find_memtables, expr.op()):
+            self._register_in_memory_table(memtable)
+
+        result = self.fetch_from_cursor(
+            self.raw_sql(sql, external_tables=external_tables),
+            table_expr.schema(),
+        )
+        result_handler = _get_result_handler(expr)
```

Comment:
`_get_result_handler` is not used anywhere else, so 
```suggestion
        if isinstance(expr, ir.Scalar):
            return result.iloc[0, 0]
        elif isinstance(expr, ir.Column):
            return result.iloc[:, 0]
        else:
            return result
```